superduperdb/base/datalayer.py

```diff
@@ -1042,23 +1044,39 @@ def replace(
             info = self.metadata.get_component(
                 object.type_id, object.identifier, version=object.version
             )
-        except FileNotFoundError:
+        except FileNotFoundError as e:
             if upsert:
                 return self.apply(
                     object,
                 )
-            raise FileNotFoundError
+            raise e
 
         # If object has no version, update the last version
         object.version = info['version']
 
+        serialized = object.dict().encode(leaves_to_keep=(Component,))
+
+        children = [
+            v for v in serialized['_leaves'].values() if isinstance(v, Component)
+        ]
+
+        for child in children:
+            self.replace(child)
+
+        for k, v in serialized['_leaves'].items():
+            if isinstance(v, Component):
+                serialized['_leaves'][
+                    k
+                ] = f'?db.load({v.type_id}, {v.identifier}, {v.version})'
+
         self.artifact_store.delete_artifact(info)
-        new_info = object.dict().encode()
-        new_info = self.artifact_store.save_artifact(new_info)
+
+        serialized = self.artifact_store.save_artifact(serialized)
```

Comment:
Should we need to remove the relationship between parent and it's old children?

Reply:
Yes good point. I'll check that.

Reply:
Yes, we need to update the links to all old children as well.

Reply:
probably:
```python
old_object = db.load(..., object.identifier)
old_object.childrens ...
```